### ISSUE TITLE
Delay when using WASD/arrow keys with NVDA and JAWS

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/balloons-and-static-electricity#409]
For https://github.com/phetsims/QA/issues/112.

From https://github.com/phetsims/balloons-and-static-electricity/issues/402, @lmulhall-phet noticed this.

There is sometimes a 0.5-2 second delay when releasing a WASD or arrow key while moving the balloon. The result is that the balloon will continue to move after a user releases a key. The delay is usually around 0.5 seconds, sometimes it is longer.

The delay seems to be much worse after a balloon picks up charges on the sweater.

My best guess is that the AT is struggling with updates to the accessibility tree since we update the DOM many many times per second. Some low hanging fruit would be to batch the updates to reduce this. It would take ~10 minutes to test this theory, then 2-4 hours to implement a fix if it turns out this is correct.

I will test. @emily-phet @terracoda how blocking is this delay, does it need to be fixed before publication?

**Comment by terracoda** [comment]
@jessegreenberg 2 seconds is kind of a long time. Do the alerts stay in synch with what is happening?

**Comment by terracoda** [comment]
@jessegreenberg, I would hope we that we could fix this before publication, but I would like to discuss with @emily-phet what kinds of issues constitute a block.

I have witnessed this type of performance delay behaviour in earlier versions, and I found it not very nice, so ideally, it would be great to fix it.

**Comment by terracoda** [comment]
@jessegreenberg, I think I need more details about the context in which performance becomes bad. 

Have you tested your theory about "AT struggling with updates to the Accessibility Tree"?

**Comment by jessegreenberg** [comment]
All updates to inner content go through AccessibiliytUtil.setTextContent  - I tried to short circuit that by returning early but I don't see much of a performance benefit or difference when I do that. It seems slightly better but I am not convinced it is enough to warrant optimizing there.

**Comment by jessegreenberg** [comment]
I tried to short circuit setLabelContent,  setDescriptionContent, setInnerContent and observed similar results.

**Comment by jessegreenberg** [comment]
I tried disabling charge animation, it didn't improve performance.

**Comment by jessegreenberg** [comment]
Are there occasional delays in keyup events in another context?

**Comment by jessegreenberg** [comment]
Not really, I tried this HTML:
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>title</title>
  </head>
  <body>

    <div role="application">
      <div id="balloon" tabindex="0">Yellow Balloon</div>
    </div>

    <p id="keydown-readout">Keydown Readout</p>
    <p id="keyup-readout" aria-live="assertive">Keyup Readout</p>

  
  </body>
  <script type="text/javascript">
    var downCount = 0;
    var upCount = 0;
    var elements = [];

    document.getElementById( "balloon" ).addEventListener( 'keydown', function() {
      document.getElementById( 'keydown-readout' ).innerHTML = "Keydown Readout: " + downCount++;

      for ( var i = 0;  i < elements.length; i++ ) {
        elements[ i ].innerHTML = 'something random - ' + Math.random();
      }
    } );

    document.getElementById( "balloon" ).addEventListener( 'keyup', function() {
      document.getElementById( 'keyup-readout' ).innerHTML = "Keyup Readout: " + upCount++;
    } );

    // create a bunch of DOM elements to update content every down
    for ( var i = 0; i < 100; i++ ) {
      var newElement = document.createElement( 'p' );
      elements.push( newElement );

      newElement.innerHTML = Math.random();
      document.body.appendChild( newElement );
    }
  </script>
</html>
```

And I did not see the same delay in key events that we are in the sim.

**Comment by jessegreenberg** [comment]
Trying performance profiling in FF with NVDA:
![profile](https://user-images.githubusercontent.com/6396244/40865287-c94581a2-65b4-11e8-9fed-ce4f95e857df.PNG)

At 1600 ms I tapped the A key and the balloon moved left until 3000 ms. Can see the framerate drop, in that time.

**Comment by jessegreenberg** [comment]
What is all that green?

**Comment by jessegreenberg** [comment]
Alright, here is a different profile:
![capture](https://user-images.githubusercontent.com/6396244/40866146-5fab703a-65b9-11e8-9e47-f15e77050d5e.PNG)

1600-4800ms I am dragging with the mouse. 8000 - 13000ms I am dragging with the keyboard. Every sharp drop is a keydown event, so we can see FF/NVDA struggles with this input more. So this makes me think it is more about the input than the PDOM updates becuase the PDOM should be updating frequently from both input.

**Comment by jessegreenberg** [comment]
![capture](https://user-images.githubusercontent.com/6396244/40866347-72d952de-65ba-11e8-94eb-54c2799d5c06.PNG)

Here is a profile when FF is turned off. Each drop in framerate is from a keydown event. So keyboard navigation is still taxing, but less so. But I don't understand why paint seems to be called more frequently when the AT is on.

**Comment by jessegreenberg** [comment]
Wow...NVDA does not have this delay problem at all in FF in the a11y view. Why on earth is that?

**Comment by terracoda** [comment]
@jessegreenberg, is the only difference the existence of an iframe?

And, What is "Paint"?

**Comment by jessegreenberg** [comment]
> @jessegreenberg, is the only difference the existence of an iframe?

I wondered about that, but other iframes (like those for phetio) do have this problem. Might be some other aspect of the a11y-view document or css settings or something?

Apparently the Paint cycle the last step the browser takes to render all pixels onto the screen. https://developers.google.com/web/fundamentals/performance/critical-rendering-path/render-tree-construction

**Comment by terracoda** [comment]
@jessegreenberg, did you mean to say "other iframes (like those for phetio) do **not** have this problem" in https://github.com/phetsims/balloons-and-static-electricity/issues/409#issuecomment-394056610.

**Regarding Paint**
From that article you posted:
> Here's a quick recap of the browser's steps:
> 1. Process HTML markup and build the DOM tree.
> 2. Process CSS markup and build the CSSOM tree.
> 3. Combine the DOM and CSSOM into a render tree.
> 4. Run layout on the render tree to compute geometry of each node.
> 5. Paint the individual nodes to the screen.

> If either the DOM or CSSOM were modified, you would have to repeat the process in order to figure out which pixels would need to be re-rendered on the screen.

> Optimizing the critical rendering path is the process of minimizing the total amount of time spent performing steps 1 through 5 in the above sequence. Doing so renders content to the screen as quickly as possible and also reduces the amount of time between screen updates after the initial render; that is, achieve higher refresh rates for interactive content.

Since we have a dynamic PDOM, is possible the browser is going through steps 1-5 each time it changes?

**Comment by terracoda** [comment]
@jessegreenberg, I understand what you wrote in #409 (comment). The A11y View does not have this issue, but other sims with iframes, do.

**Comment by jessegreenberg** [comment]
Here is what I have found this morning:
 
 - The delay is generally worse after the balloon hits the sweater/picks up a charge.
 - If I disable setting inner content of PDOM elements, all alerts, and updating `hidden` attribute on DOM elements, the problem goes away away, but returns every time the balloon picks up charges.

This still doesn't explain why it doesn't happen in the a11y-view. The framerate is much worse in the a11y-view with NVDA on though.

**Comment by jessegreenberg** [comment]
Ahh, in https://github.com/phetsims/balloons-and-static-electricity/issues/409#issuecomment-394019940, if I add a paragraph to hide/unhide on keydown/keyup
```js
    document.getElementById( "balloon" ).addEventListener( 'keydown', function() {
      document.getElementById( 'element-to-hide' ).hidden = true;
```
```js
    document.getElementById( "balloon" ).addEventListener( 'keyup', function() {
      document.getElementById( 'element-to-hide' ).hidden = false;
```
I start to see long delays on the keyup event in this basic example.

**Comment by jessegreenberg** [comment]
Also, the delay is only there when setting hidden to true AND updating the DOM frequently.

**Comment by jessegreenberg** [comment]
When I reduce the number of elements to update in https://github.com/phetsims/balloons-and-static-electricity/issues/409#issuecomment-394019940, performance is much improved. We update the inner content of PDOM elements ~200 times a second in  BASE. If we can reduce this significantly, perhaps NVDA  performance will be much improved.

**Comment by terracoda** [comment]
@jessegreenberg, do you think the updates are more when rubbing on the sweater because in the visual sim, each charge is a separate image, or drawn separately? I actually not 100% percent sure how the charges are drawn,. I am assuming they are drawn separately and each represents an individual image. When rubbing on sweater, each charge that disappears is re-drawn on the balloon.

That seems like it would be a lot of "Paint". Does it have to do with that? Again not sure why it does not happen in the A11y View.

**Comment by terracoda** [comment]
@jessegreenberg, what if you only update the inner content as fast as the sim can draw charges? Would that work? Would it be more or less frequent than it is now?

**Comment by jessegreenberg** [comment]
Exactly, batching these updates will probably be one way to improve the performance.
In the version with https://github.com/phetsims/balloons-and-static-electricity/issues/409#issuecomment-394416969, If I update the example to only update ~500 times a second, performance is OK. In BASE, we are updating far less. So this is still being caused by a collection of things.

If I short circuit `BalloonModel.step()`, the performance while dragging on the sweater and around the wall is much better with NVDA on. So I think these are the two areas we should focus our efforts.

- Performance improvements to BalloonModel.dragBalloon() - It looks like most of the work is being done in `SweaterModel.checkAndTransferCharges`.
- If things are still pretty bad, we can look into batching updates to inner DOM content (4-6 hours of work in scenery).

**Comment by jessegreenberg** [comment]
This could be pretty expensive:

```js
      // check each minus charge to see whether it should be moved to the balloon
      this.minusCharges.forEach( function( minusCharge ) {
        if ( !minusCharge.movedProperty.get() && balloonShape.containsPoint( minusCharge.location ) ) {
          self.moveChargeTo( minusCharge, balloon );
          chargeMoved = true;
        }
      } );
```

**Comment by terracoda** [comment]
Great! Thanks for the update.

**Comment by jessegreenberg** [comment]
Indeed, I notice an improvement if I don't run it. It is expensive because `balloonShape` is an ellipse.

**Comment by terracoda** [comment]
Oh, if only we had round balloons. Are circles cheaper?

**Comment by jessegreenberg** [comment]
I would expect rectangles to be faster, replacing the ellipse with rectangular bounds didn't improve things much. Diving  into `moveChargeTo`.

**Comment by jessegreenberg** [comment]
The above commit renders charge nodes in the wall with a CanvasNode, and this is giving us a huge improvement, there is no delay anymore when rubbing on the sweater. The only noticeable delay for me is on the first movement of the balloon.

**Comment by jessegreenberg** [comment]
I am satisfied with performance now, and I think we can merge with master and close.

In general, I think the performance improvements in scenery will be worthwhile but I don't think they are required prior to the publication of BASE.

I tested in JAWS and NVDA. There is no delay in JAWS, and the delay in NVDA is 0.5s or less. I don't think that will even be noticeable when using a screen reader.

Ill go ahead and merge with master.

**Comment by jessegreenberg** [comment]
Merged with master, I think this issue can be closed.
